noodles-util/src/alignment/async/io/writer.rs

```diff
@@ -0,0 +1,116 @@
+    Bam(bam::r#async::io::Writer<W>),
+    /// CRAM.
+    Cram(cram::r#async::io::Writer<W>),
+}
+
+impl<W> Writer<W>
+where
+    W: AsyncWrite + Unpin,
+{
+    /// Writes a SAM header.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() -> tokio::io::Result<()> {
+    /// use noodles_sam as sam;
+    /// use noodles_util::alignment::r#async::io::writer::Builder;
+    /// use tokio::io::{self, AsyncWriteExt};
+    ///
+    /// let mut writer = Builder::default().build_from_writer(io::sink()).await?;
+    ///
+    /// let header = sam::Header::default();
+    /// writer.write_header(&header).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn write_header(&mut self, header: &sam::Header) -> io::Result<()> {
+        match self {
+            Self::Sam(writer) => writer.write_header(header).await,
+            Self::Bam(writer) => writer.write_header(header).await,
+            Self::Cram(writer) => {
+                writer.write_file_definition().await?;
+                writer.write_file_header(header).await
+            }
+        }
+    }
+
+    /// Writes an alignment record.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() -> tokio::io::Result<()> {
+    /// use noodles_sam as sam;
+    /// use noodles_util::alignment::r#async::io::writer::Builder;
+    /// use tokio::io::{self, AsyncWriteExt};
+    ///
+    /// let mut writer = Builder::default().build_from_writer(io::sink()).await?;
+    ///
+    /// let header = sam::Header::default();
+    /// let record = sam::Record::default();
+    /// writer.write_record(&header, &record).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn write_record(
+        &mut self,
+        header: &sam::Header,
+        record: &dyn sam::alignment::Record,
+    ) -> io::Result<()> {
+        match self {
+            Self::Sam(writer) => writer.write_alignment_record(header, record).await,
+            Self::Bam(writer) => writer.write_alignment_record(header, record).await,
+            Self::Cram(writer) => {
+                let record = cram::Record::try_from_alignment_record(header, record)?;
+                writer.write_record(header, record).await
+            }
+        }
+    }
+
+    /// Shuts down the alignment format writer.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() -> tokio::io::Result<()> {
+    /// use noodles_sam as sam;
+    /// use noodles_util::alignment::{self, io::Format};
+    /// use tokio::io;
+    ///
+    /// let mut writer = alignment::r#async::io::writer::Builder::default()
+    ///     .set_format(Format::Sam)
+    ///     .build_from_writer(io::sink()).await?;
+    ///
+    /// let header = sam::Header::default();
+    /// writer.finish(&header).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn finish(&mut self, header: &sam::Header) -> io::Result<()> {
```

Comment:
> Again, the individual BAM and CRAM methods that are analogous are called `shutdown` instead of `finish`. I took the name finish to mirror the sync alignment API. Not sure if these too should be harmonised?

This is intentional. Async I/O uses the finalization term `shutdown`, e.g., [`tokio::io::AsyncWriteExt::shutdown`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.shutdown). In blocking I/O, we typically rely on `drop` for implicit finalization. For explicit finalization, we just use the same terminology as used in common writers like [`flate2::write::GzEncoder::finish`](https://docs.rs/flate2/latest/flate2/write/struct.GzEncoder.html#method.finish).

Since this is async I/O, this method should use the name `shutdown` instead of `finish`.